src/Storages/ExternalStream/Kafka/ConsumerPool.h

```diff
@@ -52,6 +52,12 @@ class ConsumerPool : public IConsumerPool, private PoolBase<Consumer>
     {
     }
 
+    ~ConsumerPool() override
+    {
+        LOG_INFO(log, "Waiting for all consumers to be freed");
+        waitForNoMoreInUse(/*timeout_ms=*/5000);
```

Comment:
don't understand why this fix the problem ? It is the pooled entry trying put back to the pool but the pool is gone ?

Reply:
And also please use clang-format to format the new code

Reply:
> The other way to fix this issues is dtor ConsumerPool last which i think is what the existing pooled entry does

Yup, however, in the cases where this failed, the `Kafka` instacne was destroy before the `KafkaSource` instances.

> If this is the case, i think we will need forever instead of timed.
That was my initial thought, but concerned about if it would blocking forever. But I agreed that it's better to keep waiting in this PR first, and if we find out that this could lead to other issues, we revisit it again.

Reply:
Can we arrange Kafka instance to get destroyed before Kafka source?